server/src/main/java/org/opensearch/search/optimization/ranges/OptimizationContext.java

```diff
@@ -0,0 +1,410 @@
+            );
+            return false;
+        }
+
+        Ranges ranges = prepareFromSegment(leafCtx);
+        if (ranges == null) return false;
+
+        aggregatorBridge.tryFastFilterAggregation(values, incrementDocCount, ranges);
+
+        optimizedSegments++;
+        logger.debug("Fast filter optimization applied to shard {} segment {}", shardId, leafCtx.ord);
+        logger.debug("crossed leaf nodes: {}, inner nodes: {}", leaf, inner);
+        return true;
+    }
+
+    /**
+     * Even when ranges cannot be built at shard level, we can still build ranges
+     * at segment level when it's functionally match-all at segment level
+     */
+    private Ranges prepareFromSegment(LeafReaderContext leafCtx) throws IOException {
+        if (!preparedAtShardLevel && !aggregatorBridge.segmentMatchAll(leafCtx)) {
+            return null;
+        }
+
+        Ranges ranges = this.ranges;
+        if (ranges == null) { // not built at shard level but segment match all
+            logger.debug("Shard {} segment {} functionally match all documents. Build the fast filter", shardId, leafCtx.ord);
+            ranges = prepare(leafCtx);
+        }
+        return ranges;
+    }
+
+    /**
+     * Internal ranges representation for the optimization
+     */
+    final static class Ranges {
+        byte[][] lowers; // inclusive
+        byte[][] uppers; // exclusive
+        int size;
+        int byteLen;
+        static ArrayUtil.ByteArrayComparator comparator;
+
+        Ranges(byte[][] lowers, byte[][] uppers) {
+            this.lowers = lowers;
+            this.uppers = uppers;
+            assert lowers.length == uppers.length;
+            this.size = lowers.length;
+            this.byteLen = lowers[0].length;
+            comparator = ArrayUtil.getUnsignedComparator(byteLen);
+        }
+
+        public int firstRangeIndex(byte[] globalMin, byte[] globalMax) {
+            if (compareByteValue(lowers[0], globalMax) > 0) {
+                return -1;
+            }
+            int i = 0;
+            while (compareByteValue(uppers[i], globalMin) <= 0) {
+                i++;
+                if (i >= size) {
+                    return -1;
+                }
+            }
+            return i;
+        }
+
+        public static int compareByteValue(byte[] value1, byte[] value2) {
+            return comparator.compare(value1, 0, value2, 0);
+        }
+
+        public static boolean withinLowerBound(byte[] value, byte[] lowerBound) {
+            return compareByteValue(value, lowerBound) >= 0;
+        }
+
+        public static boolean withinUpperBound(byte[] value, byte[] upperBound) {
+            return compareByteValue(value, upperBound) < 0;
+        }
+    }
+
+    private static class RangeCollectorForPointTree {
+        private final BiConsumer<Integer, Integer> incrementRangeDocCount;
+        private int counter = 0;
+
+        private final Ranges ranges;
+        private int activeIndex;
```

Comment:
Can the multiRangesTraverse logic be part of separate class?

Reply:
It would make sense to move the traverse logic away from context if other parts from context is shareable to other optimization like term agg. And that's what I planned before.

Did a quick try https://github.com/opensearch-project/OpenSearch/pull/14464/commits/510a052e31b22ac89efaf41e389bc2cdc50dbb89
It seems to me this is probably over-engineer. We have decent complexities in ranges rewrite optimization — build ranges, handle segment level match all, while term agg is probably just do optimization for each segment.

Reply:
Although it's over-engineer to abstract the optimization context out. multiRangesTraverse and related static methods can be easily moved to the Helper class in the same package

Reply:
All tree traversal logic is moved out